cilium/bpf_metadata.cc

```diff
@@ -208,6 +208,64 @@ uint32_t Config::resolveSourceIdentity(const PolicyInstanceConstSharedPtr policy
   return source_identity;
 }
 
+SourceIPs
+Config::getL7LBEastWestSourceIPs(const PolicyInstanceConstSharedPtr policy,
+                                 const Network::Address::Ip* sip,
+                                 const Network::Address::InstanceConstSharedPtr sourceAddress) {
+  struct SourceIPs sourceIPs;
+
+  const auto& ips = policy->getEndpointIPs();
+  switch (sip->version()) {
+
+    // Keep the original source address for the matching IP version, create a new source IP for
+    // the other version (with the same source port number) in case an upstream of a different
+    // IP version is chosen.
+  case Network::Address::IpVersion::v4:
+    sourceIPs.ipv4 = sourceAddress;
+    if (ips.ipv6_) {
+      sockaddr_in6 sa6 = *reinterpret_cast<const sockaddr_in6*>(ips.ipv6_->sockAddr());
+      sa6.sin6_port = htons(sip->port());
+      sourceIPs.ipv6 = std::make_shared<Network::Address::Ipv6Instance>(sa6);
+    }
+    break;
+  case Network::Address::IpVersion::v6:
+    sourceIPs.ipv6 = sourceAddress;
+    if (ips.ipv4_) {
+      sockaddr_in sa4 = *reinterpret_cast<const sockaddr_in*>(ips.ipv4_->sockAddr());
+      sa4.sin_port = htons(sip->port());
+      sourceIPs.ipv4 = std::make_shared<Network::Address::Ipv4Instance>(&sa4);
+    }
+    break;
+  }
+
+  return sourceIPs;
+}
```

Comment:
In general, I'd like to see factored-out functions as "generic utilities", while the business logic would be kept on the caller. Here "east/west" vs. "north/south" parlance would be part of the business logic.

To that end, consider:
- reuse `IPAddressPair` type, and take reference to it as a parameter instead of `policy` (so the `policy->getEndpointIPs()` call would be on the caller side)
- this function keeps the source address, and fills in the other address version from the given `IPAddressPair`, so maybe call this `IPAddressPair getIPAddressPairFrom(const Network::Address::Ip* sip, const IPAddressPair& addresses)`?
- add a comment above explaining the above

Reply:
thanks. changed